sycl/source/detail/jit_compiler.cpp

```diff
@@ -678,6 +688,102 @@ updatePromotedArgs(const ::jit_compiler::SYCLKernelInfo &FusedKernelInfo,
   }
 }
 
+sycl::detail::pi::PiKernel jit_compiler::materializeSpecConstants(
+    QueueImplPtr Queue, RTDeviceBinaryImage *BinImage,
+    const std::string &KernelName, std::vector<unsigned char> &SpecConstBlob) {
+  if (!BinImage) {
+    throw sycl::exception(sycl::make_error_code(sycl::errc::invalid),
+                          "No suitable IR available for materializing");
+    return nullptr;
+  }
+  if (KernelName.empty()) {
+    throw sycl::exception(
+        sycl::make_error_code(sycl::errc::invalid),
+        "Cannot jit kernel with invalid kernel function name");
+    return nullptr;
+  }
+  auto &PM = detail::ProgramManager::getInstance();
+  if (auto CachedKernel =
+          PM.getCachedMaterializedKernel(KernelName, SpecConstBlob))
+    return CachedKernel;
+
+  auto &RawDeviceImage = BinImage->getRawData();
+  auto DeviceImageSize = static_cast<size_t>(RawDeviceImage.BinaryEnd -
+                                             RawDeviceImage.BinaryStart);
+  // Set 0 as the number of address bits, because the JIT compiler can set
+  // this field based on information from SPIR-V/LLVM module's data-layout.
+  auto BinaryImageFormat = translateBinaryImageFormat(BinImage->getFormat());
+  if (BinaryImageFormat == ::jit_compiler::BinaryFormat::INVALID) {
+    throw sycl::exception(sycl::make_error_code(sycl::errc::invalid),
+                          "No suitable IR available for materializing");
+    return nullptr;
+  }
+  ::jit_compiler::SYCLKernelBinaryInfo BinInfo{
+      BinaryImageFormat, 0, RawDeviceImage.BinaryStart, DeviceImageSize};
+
+  ::jit_compiler::TargetInfo TargetInfo = getTargetInfo(Queue);
+  AddToConfigHandle(
+      ::jit_compiler::option::JITTargetInfo::set(std::move(TargetInfo)));
+  bool DebugEnabled =
+      detail::SYCLConfig<detail::SYCL_RT_WARNING_LEVEL>::get() > 0;
+  AddToConfigHandle(
+      ::jit_compiler::option::JITEnableVerbose::set(DebugEnabled));
+  AddToConfigHandle(::jit_compiler::option::JITEnableCaching::set(
+      detail::SYCLConfig<detail::SYCL_ENABLE_FUSION_CACHING>::get()));
+
+  std::string TargetCPU =
+      detail::SYCLConfig<detail::SYCL_JIT_TARGET_CPU>::get();
+  std::string TargetFeatures =
+      detail::SYCLConfig<detail::SYCL_JIT_TARGET_FEATURES>::get();
```

Comment:
Would it make sense to turn these into options, too? This way, the kernel fusion entry point could also use them to produce binaries tailored towards the architecture given via environment variables.

Reply:
So that we still have an env variables, but they get stored as [`option`](https://github.com/jchlanda/llvm/blob/jakub/jit_spec_const/sycl-fusion/jit-compiler/include/Options.h#L76) and accessed through [`ConfigHelper`](https://github.com/jchlanda/llvm/blob/jakub/jit_spec_const/sycl-fusion/jit-compiler/lib/helper/ConfigHelper.h#L49)?

If that's it, I'm cool with that, but would prefer to do it as a follow up, to not get distracted from this (already large) PR.

Reply:
Yes, exactly. We would keep the environment variables, evaluate them here in the SYCL RT and pass them through the `ConfigHelper`, so both entry points to the JIT have access to the information without having to add two more `const char*` parameters to each entry point. 

I'm fine with doing that in a follow-up PR, no problem.

Reply:
Cool, I've created an internal ticket for it (alongside the rename), so it doesn't get lost.